File: src/schema.rs
//! Schema generation and validation for Conclaude configuration files.
//!
//! This module provides utilities for generating JSON Schema definitions from the
//! `ConclaudeConfig` structure, writing schemas to files, and validating YAML
//! configuration files against the schema.
//!
//! # Primary Use Cases
//!
//! 1. **Build/Release Automation**: The `generate-schema` binary uses these functions
//!    to create schema files during the build process
//! 2. **GitHub Releases**: Generated schemas are automatically uploaded to GitHub
//!    releases for public consumption
//! 3. **IDE Integration**: YAML language server headers reference the published schema
//!    for autocompletion and validation in editors
//!
//! # Schema Distribution
//!
//! The generated schema is published at:
//! `https://github.com/connix-io/conclaude/releases/latest/download/conclaude-schema.json`
//!
//! This URL is embedded in configuration files via the `yaml-language-server` comment
//! header, enabling IDE support for `.conclaude.yml` files.
//!
//! # Examples
//!
//! ```rust
//! use conclaude::schema::{generate_config_schema, write_schema_to_file, validate_config_against_schema};
//! use std::path::PathBuf;
//!
//! // Generate a schema
//! let schema = generate_config_schema().unwrap();
//!
//! // Write to file
//! let path = PathBuf::from("schema.json");
//! write_schema_to_file(&schema, &path).unwrap();
//!
//! // Validate configuration
//! let config_yaml = r#"
//! stop:
//!   commands:
//!     - run: "cargo test"
//!   infinite: false
//! "#;
//! validate_config_against_schema(config_yaml).unwrap();
//! ```

use anyhow::{Context, Result};
use schemars::schema_for;
use serde_json::{json, Value};
use std::fs;
use std::path::PathBuf;

use crate::config::ConclaudeConfig;

/// Generates a JSON Schema for the `ConclaudeConfig` structure
///
/// # Errors
///
/// Returns an error if JSON schema generation or value serialization fails.
#[allow(dead_code)]
pub fn generate_config_schema() -> Result<Value> {
    let schema = schema_for!(ConclaudeConfig);

    // Enhance the schema with additional metadata
    let mut schema_value = serde_json::to_value(schema)?;

    // Add metadata to the root schema
    if let Value::Object(ref mut schema_obj) = schema_value {
        schema_obj.insert(
            "$schema".to_string(),
            json!("http://json-schema.org/draft-07/schema#"),
        );
        schema_obj.insert("title".to_string(), json!("Conclaude Configuration"));
        schema_obj.insert(
            "description".to_string(),
            json!("Configuration schema for Conclaude - Claude Code hook handler"),
        );
        schema_obj.insert("$id".to_string(), json!("https://github.com/connix-io/conclaude/releases/latest/download/conclaude-schema.json"));
    }

    Ok(schema_value)
}

/// Writes the generated schema to a file
///
/// # Errors
///
/// Returns an error if JSON serialization fails, directory creation fails, or file writing fails.
#[allow(dead_code)]
pub fn write_schema_to_file(schema: &Value, output_path: &PathBuf) -> Result<()> {
    let schema_json =
        serde_json::to_string_pretty(schema).context("Failed to serialize schema to JSON")?;

    // Ensure parent directory exists
    if let Some(parent) = output_path.parent() {
        fs::create_dir_all(parent)
            .with_context(|| format!("Failed to create directory: {}", parent.display()))?;
    }

    fs::write(output_path, schema_json)
        .with_context(|| format!("Failed to write schema to: {}", output_path.display()))?;

    Ok(())
}

/// Validates that a given YAML content matches the schema
///
/// # Errors
///
/// Returns an error if the YAML content is invalid or does not match the expected structure.
#[allow(dead_code)]
pub fn validate_config_against_schema(config_content: &str) -> Result<()> {
    // Parse the YAML to ensure it's valid
    let _: ConclaudeConfig = serde_yaml::from_str(config_content).map_err(|e| {
        let base_error = e.to_string();
        let mut parts = vec![
            "Configuration validation failed".to_string(),
            String::new(),
            format!("Error: {}", base_error),
        ];

        // Add specific guidance based on error type
        if base_error.contains("unknown field") {
            parts.push(String::new());
            parts.push("The configuration contains an unknown field.".to_string());
            parts.push("Check the field name for typos or incorrect casing.".to_string());
        } else if base_error.contains("invalid type") {
            parts.push(String::new());
            parts.push("A field has the wrong type (e.g., string instead of boolean).".to_string());
        } else if base_error.contains("expected") || base_error.contains("while parsing") {
            parts.push(String::new());
            parts.push("YAML syntax error detected. Check indentation and formatting.".to_string());
        }

        anyhow::anyhow!(parts.join("\n"))
    })?;

    Ok(())
}

/// Gets the default schema URL for YAML language server headers
#[must_use]
pub fn get_schema_url() -> String {
    "https://github.com/connix-io/conclaude/releases/latest/download/conclaude-schema.json"
        .to_string()
}

/// Generates a YAML language server header comment with schema URL
#[must_use]
pub fn generate_yaml_language_server_header(custom_schema_url: Option<&str>) -> String {
    let default_url = get_schema_url();
    let schema_url = custom_schema_url.unwrap_or(&default_url);
    format!("# yaml-language-server: $schema={schema_url}\n")
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_generate_config_schema() {
        let schema = generate_config_schema().unwrap();

        // Check that the schema is a valid JSON object
        assert!(schema.is_object());

        // Check for required metadata
        let schema_obj = schema.as_object().unwrap();
        assert!(schema_obj.contains_key("$schema"));
        assert!(schema_obj.contains_key("title"));
        assert!(schema_obj.contains_key("description"));
        assert!(schema_obj.contains_key("$id"));
        assert!(schema_obj.contains_key("type"));
        assert!(schema_obj.contains_key("properties"));
    }

    #[test]
    fn test_write_schema_to_file() {
        let schema = generate_config_schema().unwrap();
        let temp_dir = tempdir().unwrap();
        let schema_path = temp_dir.path().join("test-schema.json");

        write_schema_to_file(&schema, &schema_path).unwrap();

        // Verify file exists and contains valid JSON
        assert!(schema_path.exists());
        let content = fs::read_to_string(&schema_path).unwrap();
        let _: Value = serde_json::from_str(&content).unwrap();
    }

    #[test]
    fn test_validate_config_against_schema() {
        let valid_config = r#"
stop:
  commands:
    - run: "echo test"
  infinite: false
preToolUse:
  preventRootAdditions: true
  uneditableFiles: []
"#;

        validate_config_against_schema(valid_config).unwrap();

        let invalid_config = r#"
invalid_field: "should fail"
"#;

        assert!(validate_config_against_schema(invalid_config).is_err());
    }

    #[test]
    fn test_generate_yaml_language_server_header() {
        let default_header = generate_yaml_language_server_header(None);
        assert!(default_header.contains("yaml-language-server:"));
        assert!(default_header.contains("github.com/connix-io/conclaude"));

        let custom_header =
            generate_yaml_language_server_header(Some("https://example.com/schema.json"));
        assert!(custom_header.contains("https://example.com/schema.json"));
    }

    #[test]
    fn test_get_schema_url() {
        let url = get_schema_url();
        assert!(url.starts_with("https://"));
        assert!(url.contains("github.com"));
        assert!(url.to_lowercase().ends_with(".json"));
    }
}

File: internal/schema/schema.go
package schema

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/connix-io/conclaude/internal/config"
	"github.com/invopop/jsonschema"
	jsonschemavalidator "github.com/santhosh-tekuri/jsonschema/v6"
	"gopkg.in/yaml.v3"
)

const (
	DefaultSchemaURL = "https://raw.githubusercontent.com/connix-io/conclaude/" +
		"main/conclaude-schema.json"

	// File permission constants
	SchemaFilePermission = 0600

	// String constants
	EmptyString = ""
)

// GenerateSchema generates JSON schema for the ConclaudeConfig.
func GenerateSchema() (string, error) {
	reflector := jsonschema.Reflector{
		AllowAdditionalProperties:  false,
		RequiredFromJSONSchemaTags: true,
		DoNotReference:             true,
	}

	schema := reflector.Reflect(&config.ConclaudeConfig{})

	// Set schema metadata
	schema.Title = "Conclaude Configuration"
	schema.Description = "Configuration schema for conclaude hook handler"
	// Remove the Version field - let the library handle it

	schemaBytes, err := json.MarshalIndent(schema, "", "  ")
	if err != nil {
		return "", fmt.Errorf("failed to marshal schema to JSON: %w", err)
	}

	return string(schemaBytes), nil
}

// WriteSchemaToFile writes the JSON schema to a file.
func WriteSchemaToFile(schema string, filePath string) error {
	err := os.WriteFile(filePath, []byte(schema), SchemaFilePermission)
	if err != nil {
		return fmt.Errorf(
			"failed to write schema to file %s: %w",
			filePath,
			err,
		)
	}

	return nil
}

// ValidateConfigAgainstSchema validates a YAML config string against the
// JSON schema.
func ValidateConfigAgainstSchema(configYAML string) error {
	// First, convert YAML to JSON for validation
	var yamlData interface{}
	err := yaml.Unmarshal([]byte(configYAML), &yamlData)
	if err != nil {
		return fmt.Errorf("failed to parse YAML: %w", err)
	}

	jsonData, err := json.Marshal(yamlData)
	if err != nil {
		return fmt.Errorf("failed to convert YAML to JSON: %w", err)
	}

	// Generate the schema
	schemaStr, err := GenerateSchema()
	if err != nil {
		return fmt.Errorf("failed to generate schema: %w", err)
	}

	// Parse schema to add as resource
	var schemaDoc interface{}
	if err := json.Unmarshal([]byte(schemaStr), &schemaDoc); err != nil {
		return fmt.Errorf("failed to parse schema: %w", err)
	}
	
	// Compile the schema using the v6 API
	compiler := jsonschemavalidator.NewCompiler()
	compiler.DefaultDraft(jsonschemavalidator.Draft7)
	
	// Add the schema as a resource
	if err := compiler.AddResource("schema.json", schemaDoc); err != nil {
		return fmt.Errorf("failed to add schema resource: %w", err)
	}
	
	// Compile the schema
	schema, err := compiler.Compile("schema.json")
	if err != nil {
		return fmt.Errorf("failed to compile schema: %w", err)
	}

	// Parse the JSON data for validation
	var instance interface{}
	if err := json.Unmarshal(jsonData, &instance); err != nil {
		return fmt.Errorf("failed to parse JSON for validation: %w", err)
	}

	// Validate against the schema
	if err := schema.Validate(instance); err != nil {
		return fmt.Errorf("validation failed: %w", err)
	}

	return nil
}

// GenerateYAMLLanguageServerHeader generates the YAML language server header.
func GenerateYAMLLanguageServerHeader(customSchemaURL *string) string {
	schemaURL := DefaultSchemaURL
	if customSchemaURL != nil && *customSchemaURL != EmptyString {
		schemaURL = *customSchemaURL
	}

	return fmt.Sprintf("# yaml-language-server: $schema=%s\n\n", schemaURL)
}

// GetSchemaURL returns the default schema URL.
func GetSchemaURL() string {
	return DefaultSchemaURL
}
